File: day16/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"regexp"
	"strconv"
	"strings"
)

const (
	puzzleInput = "input.txt"
)

type (
	Rule struct {
		n1 int
		n2 int
		n3 int
		n4 int
	}
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatal(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatal(err)
		}
	}()

	re := regexp.MustCompile(`^([a-z ]+): ([0-9]+)-([0-9]+) or ([0-9]+)-([0-9]+)$`)

	rules := map[string]Rule{}
	ruleNames := map[string]struct{}{}

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		if line == "" {
			break
		}
		m := re.FindStringSubmatch(line)
		if m == nil {
			log.Fatal("Invalid rule line")
		}
		text := m[1]
		n1, err := strconv.Atoi(m[2])
		if err != nil {
			log.Fatal(err)
		}
		n2, err := strconv.Atoi(m[3])
		if err != nil {
			log.Fatal(err)
		}
		n3, err := strconv.Atoi(m[4])
		if err != nil {
			log.Fatal(err)
		}
		n4, err := strconv.Atoi(m[5])
		if err != nil {
			log.Fatal(err)
		}
		rules[text] = Rule{
			n1: n1,
			n2: n2,
			n3: n3,
			n4: n4,
		}
		ruleNames[text] = struct{}{}
	}
	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	ownTicket := []int{}

	for scanner.Scan() {
		line := scanner.Text()
		if line == "" {
			break
		}
		if line == "your ticket:" {
			continue
		}
		for _, i := range strings.Split(line, ",") {
			num, err := strconv.Atoi(i)
			if err != nil {
				log.Fatal("Invalid ticket format")
			}
			ownTicket = append(ownTicket, num)
		}
	}
	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	otherTickets := [][]int{}

	part1 := 0

	for scanner.Scan() {
		line := scanner.Text()
		if line == "nearby tickets:" {
			continue
		}
		ticket := []int{}
		for _, i := range strings.Split(line, ",") {
			num, err := strconv.Atoi(i)
			if err != nil {
				log.Fatal("Invalid ticket format")
			}
			ticket = append(ticket, num)
		}
		if i, ok := isInvalid(ticket, rules); ok {
			part1 += i
		} else {
			otherTickets = append(otherTickets, ticket)
		}
	}
	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	fmt.Println("Part 1:", part1)

	possible := make([]map[string]struct{}, 0, len(ownTicket))
	for i := 0; i < len(ownTicket); i++ {
		possible = append(possible, copyMap(ruleNames))
	}

	determined := map[string]int{}

	for {
		changed := false
		for n, i := range possible {
			if len(i) < 2 {
				continue
			}
			for _, j := range otherTickets {
				rm := []string{}
				for k := range i {
					if notInRange(j[n], rules[k]) {
						rm = append(rm, k)
					}
				}
				for _, r := range rm {
					delete(i, r)
					changed = true
				}
			}
		}
		if !changed {
			break
		}
	}
	for {
		changed := false
		for n, i := range possible {
			if len(i) == 0 {
				log.Fatal("Invalid constraints")
			}
			if len(i) == 1 {
				for k := range i {
					determined[k] = n
				}
			}
		}
		for k, v := range determined {
			for n, i := range possible {
				if n == v {
					continue
				}
				if _, ok := i[k]; ok {
					delete(i, k)
					changed = true
				}
			}
		}
		if !changed {
			break
		}
	}
	if len(determined) < len(ownTicket) {
		log.Fatal("Invalid constraints")
	}

	part2 := 1
	for k, v := range determined {
		if startsWith(k, "departure") {
			part2 *= ownTicket[v]
		}
	}
	fmt.Println("Part 2:", part2)
}

func isInvalid(ticket []int, rules map[string]Rule) (int, bool) {
outer:
	for _, i := range ticket {
		for _, v := range rules {
			if inRange(i, v) {
				continue outer
			}
		}
		return i, true
	}
	return 0, false
}

func inRange(i int, r Rule) bool {
	return i >= r.n1 && i <= r.n4 && (i <= r.n2 || i >= r.n3)
}

func notInRange(i int, r Rule) bool {
	return i < r.n1 || i > r.n4 || (i > r.n2 && i < r.n3)
}

func copyMap(m map[string]struct{}) map[string]struct{} {
	other := make(map[string]struct{}, len(m))
	for k, v := range m {
		other[k] = v
	}
	return other
}

func startsWith(s, p string) bool {
	return len(s) >= len(p) && s[:len(p)] == p
}

File: day16/src/main.rs
use regex::Regex;
use std::collections::{HashMap, HashSet};
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

type BErr = Box<dyn std::error::Error>;

struct Rule(usize, usize, usize, usize);

fn main() -> Result<(), BErr> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);
    let mut lines = reader.lines();

    let re = Regex::new(r"^([a-z ]+): ([0-9]+)-([0-9]+) or ([0-9]+)-([0-9]+)$")?;

    let (rules, rule_names) = {
        let mut rules = HashMap::new();
        let mut rule_names = HashSet::new();

        while let Some(line) = lines.next() {
            let line = line?;
            if line == "" {
                break;
            }
            let m = re.captures(&line).ok_or("Invalid rule line")?;
            let text = m[1].to_string();
            let text2 = m[1].to_string();
            let n1 = m[2].parse::<usize>()?;
            let n2 = m[3].parse::<usize>()?;
            let n3 = m[4].parse::<usize>()?;
            let n4 = m[5].parse::<usize>()?;
            rules.insert(text, Rule(n1, n2, n3, n4));
            rule_names.insert(text2);
        }
        (rules, rule_names)
    };

    let own_ticket = {
        let mut own_ticket = Vec::new();
        while let Some(line) = lines.next() {
            let line = line?;
            if line == "" {
                break;
            }
            if line == "your ticket:" {
                continue;
            }
            for i in line.split(",") {
                own_ticket.push(i.parse::<usize>()?);
            }
        }
        own_ticket
    };

    let (other_tickets, part1) = {
        let mut other_tickets = Vec::new();
        let mut part1 = 0;
        while let Some(line) = lines.next() {
            let line = line?;
            if line == "nearby tickets:" {
                continue;
            }
            let mut ticket = Vec::new();
            for i in line.split(",") {
                ticket.push(i.parse::<usize>()?);
            }
            if let Some(i) = is_invalid(&ticket, &rules) {
                part1 += i;
            } else {
                other_tickets.push(ticket);
            }
        }
        (other_tickets, part1)
    };
    println!("Part 1: {}", part1);

    let mut possible = (0..own_ticket.len())
        .map(|_| rule_names.clone())
        .collect::<Vec<_>>();
    let mut determined: HashMap<String, usize> = HashMap::new();
    while determined.len() < own_ticket.len() {
        loop {
            let mut changed = false;
            possible = possible
                .into_iter()
                .enumerate()
                .map(|(n, mut i)| {
                    if i.len() < 2 {
                        return i;
                    }
                    for j in &other_tickets {
                        let l = i.len();
                        i = i
                            .into_iter()
                            .filter(|k| in_range(j[n], &rules[k]))
                            .collect::<HashSet<_>>();
                        if i.len() != l {
                            changed = true;
                        }
                    }
                    i
                })
                .collect();
            if !changed {
                break;
            }
        }
        for (n, i) in possible.iter().enumerate() {
            if i.len() == 0 {
                return Err("Invalid constraints".into());
            }
            if i.len() == 1 {
                for k in i {
                    determined.insert(k.clone(), n);
                }
            }
        }
        for (k, &v) in &determined {
            possible = possible
                .into_iter()
                .enumerate()
                .map(|(n, mut i)| {
                    if n == v {
                        return i;
                    }
                    i.remove(k);
                    i
                })
                .collect();
        }
    }

    let part2 = determined
        .iter()
        .filter(|(k, _)| k.starts_with("departure"))
        .fold(1, |acc, (_, &v)| acc * own_ticket[v]);
    println!("Part 2: {}", part2);
    Ok(())
}

fn is_invalid(ticket: &[usize], rules: &HashMap<String, Rule>) -> Option<usize> {
    'outer: for &i in ticket {
        for v in rules.values() {
            if in_range(i, v) {
                continue 'outer;
            }
        }
        return Some(i);
    }
    None
}

fn in_range(i: usize, r: &Rule) -> bool {
    i >= r.0 && i <= r.3 && (i <= r.1 || i >= r.2)
}
